### PR TITLE
[feature](running_difference) support running_difference function

### PR DESCRIPTION
**Pull request by ZhangYiXi-dev**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Calculates the difference between successive row values ​​in the data block. Returns 0 for the first row and the difference from the previous row for each subsequent row.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.34 seconds
 load time: 589 seconds
 storage size: 17154821198 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027151845_clickbench_pr_34884.html

File: be/src/vec/functions/function_running_difference.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace doris::vectorized {
+
+void register_function_running_difference(SimpleFunctionFactory& factory) {
+    //factory.register_function<FunctionRunningDifferenceImpl<false>>();
+    factory.register_function<FunctionRunningDifference>();
+    //factory.register_function<FunctionCase<false, true>>();
+    //factory.register_function<FunctionCase<true, false>>();
+    //factory.register_function<FunctionCase<true, true>>();
+}
```
Comment (Yukang-Lian) [inline be/src/vec/functions/function_running_difference.cpp:11]: plz remove redundant code.

File: be/src/vec/functions/function_running_difference.h [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    bool is_deterministic() const override { return false; }
+    bool is_deterministic_in_scope_of_query() const override
+    {
+         return false;
+    } 
+
+   // bool is_suitable_for_short_circuit_arguments_execution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+    bool use_default_implementation_for_nulls() const override { return false; }
+
```
Comment (zhangstar333) [inline be/src/vec/functions/function_running_difference.h:132]: maybe it's better to return true

File: be/test/vec/function/function_running_difference_test.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    DataSet data_set = {
+                        {{(int64_t)1}, (int64_t)0},
+                        {{(int64_t)2}, (int64_t)1},
+                        {{(int64_t)3}, (int64_t)1},
+                        {{(int64_t)5}, (int64_t)2}};
+
+    check_function<DataTypeInt64, true>(func_name, input_types, data_set);
+}
+
```
Comment (Yukang-Lian) [inline be/test/vec/function/function_running_difference_test.cpp:25]: We need to add more ut to test different data types such as `float` and `date`.

File: be/src/vec/functions/function_running_difference.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        DataTypePtr res;
+        dispatchForSourceType(*remove_nullable(arguments[0]), [&](auto field_type_tag)
+        {
+            res = std::make_shared<DataTypeNumber<DstFieldType<decltype(field_type_tag)>>>();
+        });
+
+        if (arguments[0]->is_nullable())
+            res = make_nullable(res);
+
```
Comment (zhangstar333) [inline be/src/vec/functions/function_running_difference.h:143]: I see in ...functions.py file, you set return type is ALWAYS_NOT_NULLABLE, but here make_nullable

File: gensrc/script/doris_builtins_functions.py [hunk 5]
```diff
@@ -2180,8 +2180,14 @@
...
         '_ZN5doris15StringFunctions10split_partEPN9doris_udf15FunctionContextERKNS1_9StringValES6_RKNS1_6IntValE',
         '', '', 'vec', 'ALWAYS_NULLABLE'],
+
      [['extract_url_parameter'], 'VARCHAR', ['VARCHAR', 'VARCHAR'],'','', '', 'vec', ''],
 
+    # runningdifference
+    [['running_difference'], 'INT', ['INT'],
+        '',
+        '', '', 'vec', 'ALWAYS_NOT_NULLABLE'],
+
```
Comment (Yukang-Lian) [inline gensrc/script/doris_builtins_functions.py:2190]: we need to register other data types. As follows
```suggestion
[['running_difference'], 'FLOAT', ['FLOAT'],
        '',
        '', '', 'vec', 'ALWAYS_NOT_NULLABLE'],```

**Review by Yukang-Lian — CHANGES_REQUESTED** [review]
plz add regression-test and docs.

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.02 seconds
 load time: 555 seconds
 storage size: 17154699331 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028030844_clickbench_pr_35046.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.77 seconds
 load time: 559 seconds
 storage size: 17154644859 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028050643_clickbench_pr_35136.html

File: be/src/vec/functions/function_running_difference.cpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#include "vec/functions/function_running_difference.h"
+
+namespace doris::vectorized {
+
+void register_function_running_difference(SimpleFunctionFactory& factory) {
+    //factory.register_function<FunctionRunningDifferenceImpl<false>>();
+    factory.register_function<FunctionRunningDifference>();
+    //factory.register_function<FunctionCase<false, true>>();
+    //factory.register_function<FunctionCase<true, false>>();
+    //factory.register_function<FunctionCase<true, true>>();
```
Comment (dataroaring) [inline be/src/vec/functions/function_running_difference.cpp:10]: Please remove useless comment.

File: regression-test/suites/query_p0/sql_functions/math_functions/test_running_difference.sql [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+DUPLICATE KEY(id) 
+DISTRIBUTED BY HASH(id) BUCKETS 3 
+PROPERTIES ( 
+    "replication_num" = "1"
+); 
+
+INSERT into running_difference_test (id,day, time_val,doublenum) values ('1', '2022-11-08', '2022-03-12 11:05:04', 4.7),
+                                                   ('1', '2022-10-31', '2022-03-12 10:42:01', 3.3),
+                                                    ('1', '2022-10-27', '2022-03-12 10:41:02', 2.6),
+                                                   ('1', '2022-10-28', '2022-03-12 10:41:00', 1.4);
```
Comment (dataroaring) [inline regression-test/suites/query_p0/sql_functions/math_functions/test_running_difference.sql:18]: add null for test.

File: regression-test/suites/query_p0/sql_functions/math_functions/test_running_difference.sql [hunk 8]
```diff
@@ -0,0 +1,55 @@
+DROP TABLE IF EXISTS running_difference_test;
+
+CREATE TABLE running_difference_test (
+                 `id` int NULL COMMENT 'id' ,
+                `day` date COMMENT 'day', 
+	`time_val` datetime COMMENT 'time_val',
+ 	`doublenum` double NULL COMMENT 'doublenum'
+                )
```
Comment (dataroaring) [inline regression-test/suites/query_p0/sql_functions/math_functions/test_running_difference.sql:8]: add a column not nullale.

File: be/src/vec/functions/function_running_difference.h [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    bool use_default_implementation_for_nulls() const override { return false; }
+
+    bool use_default_implementation_for_constants() const override { return true; }
+
+    template <typename SrcFieldType>
+    using DstFieldType = typename NumberTraits::ResultOfSubtraction<SrcFieldType, SrcFieldType>::Type;
```
Comment (zhangstar333) [inline be/src/vec/functions/function_running_difference.h:39]: now seems it's could remove

**Review by Yukang-Lian — DISMISSED** [review]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

File: be/src/vec/functions/function_running_difference.h [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    static FunctionPtr create() { return std::make_shared<FunctionRunningDifference>(); }
+
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    bool use_default_implementation_for_nulls() const override { return false; }
+
+    bool use_default_implementation_for_constants() const override { return true; }
```
Comment (HappenLee) [inline be/src/vec/functions/function_running_difference.h:53]: `use_default_implementation_for_constants` should be false.

**Pull request merged**